crates/networking/p2p/rlpx/frame.rs

```diff
@@ -1,36 +1,248 @@
+            temp_ingress_mac.clone().finalize()[..16]
+                .try_into()
+                .map_err(|_| RLPxError::CryptographyError("Invalid header mac".to_owned()))?,
+        );
+
+        assert_eq!(header_mac, expected_header_mac.0);
+
+        let header_text = header_ciphertext;
+        // Use temporary value as it can be discarded if the buffer does not contain yet the full message
+        let mut temp_ingress_aes = self.ingress_aes.clone();
+        temp_ingress_aes.apply_keystream(header_text);
+
+        // header-data = [capability-id, context-id]
+        // Both are unused, and always zero
+        assert_eq!(&header_text[3..6], &(0_u8, 0_u8).encode_to_vec());
+
+        let frame_size: usize =
+            u32::from_be_bytes([0, header_text[0], header_text[1], header_text[2]])
+                .try_into()
+                .map_err(|_| RLPxError::CryptographyError("Invalid frame size".to_owned()))?;
+
+        let padded_size = frame_size.next_multiple_of(16);
+
+        // Check that the size is not too large to avoid a denial of
+        // service attack where the server runs out of memory.
+        if padded_size > MAX_MESSAGE_SIZE {
+            return Err(RLPxError::InvalidMessageLength());
+        }
+
+        if src.len() < 32 + padded_size + 16 {
+            // The full string has not yet arrived.
+            //
+            // We reserve more space in the buffer. This is not strictly
+            // necessary, but is a good idea performance-wise.
+            src.reserve(32 + padded_size - src.len());
+
+            // We inform the Framed that we need more bytes to form the next
+            // frame.
+            return Ok(None);
+        }
+
+        // Use advance to modify src such that it no longer contains
+        // this frame.
+        let mut frame_data = src[32..32 + padded_size + 16].to_vec();
+        src.advance(32 + padded_size + 16);
+
+        // The buffer contains the full message and will be consumed; update the ingress_mac and aes values
+        self.ingress_mac = temp_ingress_mac;
+        self.ingress_aes = temp_ingress_aes;
+
+        let (frame_ciphertext, frame_mac) = frame_data.split_at_mut(padded_size);
+
+        // check MAC
+        self.ingress_mac.update(&frame_ciphertext);
+        let frame_mac_seed = {
+            let mac_digest: [u8; 16] = self.ingress_mac.clone().finalize()[..16]
+                .try_into()
+                .map_err(|_| RLPxError::CryptographyError("Invalid mac digest".to_owned()))?;
+            let mut seed = mac_digest.into();
+            mac_aes_cipher.encrypt_block(&mut seed);
+            (H128(seed.into()) ^ H128(mac_digest)).0
+        };
+        self.ingress_mac.update(frame_mac_seed);
+        let expected_frame_mac: [u8; 16] = self.ingress_mac.clone().finalize()[..16]
+            .try_into()
+            .map_err(|_| RLPxError::CryptographyError("Invalid frame mac".to_owned()))?;
+
+        assert_eq!(frame_mac, expected_frame_mac);
```

Comment:
Not for this PR, but should we remove this and return and error? Can we have a TODO for this?